NIFI-9374 Add Standard Deprecation Logger

exceptionfactory opened the pull request:
# Summary

[NIFI-9374](https://issues.apache.org/jira/browse/NIFI-9374) Adds a `DeprecationLogger` interface with standard implementation and factory to provide a common approach to logging deprecated components and features.

The `nifi-deprecation-log` module has a single dependency on the SLF4J API to provide the logging implementation.

The `DeprecationLogger` interface has a single `warn()` method that should be sufficient for general use across components. The `DeprecationLoggerFactory` and `StandardDeprecationLogger` implementation prefix logger names with `deprecation`, enabling discrete logging to a separate appender.

The introduction of a new `nifi-deprecation.log` enables a standard approach for administrators to determine whether an installation of Apache NiFi is using components or features that will be removed in future releases. Log messages should indicate deprecated properties or capabilities, and a standard internal `DeprecationException` provides a reference to the classes and methods calling the deprecated elements.

The initial set of changes includes updates to `NiFiLegacyCipherProvider` providing example usage. The Cipher Provider supports the legacy Sensitive Properties Algorithm configuration values from Apache NiFi 1.13.2 and earlier. Deprecation logs can be observed by configuring a legacy algorithm such as `PBEWITHMD5AND256BITAES-CBC-OPENSSL`.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [X] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [X] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [X] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [X] Pull Request based on current revision of the `main` branch
- [X] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [X] Build completed using `mvn clean install -P contrib-check`
  - [X] JDK 8
  - [ ] JDK 11
  - [ ] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

Lehel44 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/conf/logback.xml:
```diff
@@ -88,6 +88,17 @@
+            <maxHistory>30</maxHistory>
+        </rollingPolicy>
+        <encoder class="ch.qos.logback.classic.encoder.PatternLayoutEncoder">
+            <pattern>%date %level [%thread] %logger %msg%n</pattern>
```
Do you think the logger name could be shortened in the layout pattern with an upper limit as in other examples?
The abbreviation algorithm with examples can be found [here](https://logback.qos.ch/manual/layouts.html#conversionWord).

exceptionfactory replied:
The logger name is not truncated in this specific scenario because it is important to know the exact class name containing deprecated references.

Under optimal circumstances, the nifi-deprecation.log will be empty, but when a deployment includes deprecated components or features, the logger name and associated stack trace will provide a reference to component usage that needs to be changed.

greyp9 commented:
Do you envision any potential for documentation impact from this changeset?  Nothing specific in mind, but I did a search to generate some ideas:

- https://docs.oracle.com/javase/7/docs/technotes/guides/javadoc/deprecation/deprecation.html
- https://kubernetes.io/docs/reference/using-api/deprecation-guide/
- https://cwiki.apache.org/confluence/display/NIFI/Release+Notes#ReleaseNotes-Version1.17.0
- https://www.javadoc.io/doc/org.apache.nifi/nifi-api/latest/deprecated-list.html (auto-generated by build)

exceptionfactory commented:
Thanks for the feedback @greyp9! Following this pull request, I anticipate additional work to integrate the `DeprecationLogger` in various components.  That should also include updates to the documentation, which will describe the `nifi-deprecation.log` with some explanation of how to evaluate steps for upgrading to current components and features.

greyp9 reviewed (commented):
Performed local runtime testing with this changeset.  I configured GenerateFlowFile -> EncryptContent(encrypt) -> EncryptContent(decrypt) -> ModifyBytes.

```
Property                    Value 
Mode                        Encrypt/Decrypt
Key Derivation Function     NiFi Legacy KDF
Encryption Algorithm        MD5_256AES
...
```
Set GFF to fire off a flowfile every five seconds.

This caused nifi-deprecation.log (and nifi-app.log) to grow at a rate of ~1MB per 8 minute interval.  It looks like the deprecation stack trace is being dumped to both log output streams.  We may want to use the `additivity` logback attribute to filter this output from app.log, like this other example:

`<logger name="org.apache.nifi.web.security.requests" level="INFO" additivity="false">`

We may also consider a strategy to log the usage only once, to filter duplicates over some time range, or maybe to log the exception less frequently.  Here's a recent PR I reviewed that attempted to address a similar problem:

https://github.com/apache/nifi/pull/6185 [NIFI-10199]

exceptionfactory commented:
Thanks for the feedback regarding the Logback configuration @greyp9!

I updated the Logback configuration to set `additivity="false"` for the `deprecation` logger to avoid duplicating messages in `nifi-app.log`. I also updated the `DEPRECATION_FILE` appender to set a maximum file size of 10 MB and a maximum history of 10. Although the log could produce many duplicate messages, this configuration limits the amount of space that could be consumed in the default configuration.

greyp9 commented:
> Thanks for the feedback regarding the Logback configuration @greyp9!
> 
> I updated the Logback configuration to set `additivity="false"` for the `deprecation` logger to avoid duplicating messages in `nifi-app.log`. I also updated the `DEPRECATION_FILE` appender to set a maximum file size of 10 MB and a maximum history of 10. Although the log could produce many duplicate messages, this configuration limits the amount of space that could be consumed in the default configuration.

Tried this out again.  The additional config does not seem to be sufficient to limit the deprecation logger output.  This configuration seems to achieve the intent:

```
        <rollingPolicy class="ch.qos.logback.core.rolling.SizeAndTimeBasedRollingPolicy">
            <fileNamePattern>${org.apache.nifi.bootstrap.config.log.dir}/nifi-deprecation_%d.%i.log</fileNamePattern>
            <maxFileSize>10MB</maxFileSize>
            <maxHistory>10</maxHistory>
            <totalSizeCap>100MB</totalSizeCap>
        </rollingPolicy>
```

This may be a corner case, due to logback being fed logging events so quickly.

exceptionfactory commented:
Thanks for the thorough testing @greyp9, great recommendation on the `totalSizeCap` property! I pushed an update add the property with a setting of `100MB`.

greyp9 reviewed (approved):
Cool, thanks!  Changes look good, and verified behavior via runtime testing.
